City-of-Helsinki/drupal-hdbt
# UHF-6198: Cache mobilemenu response to localstorage

# [UHF-6198 Mobile menu caching](https://helsinkisolutionoffice.atlassian.net/browse/UHF-6198)
Cache the mobile menu request to local storage, eliminating the need to make the same request in succession.

## What was done
Added logic to cache the mobile menu request into localstorage. Added timestamp to act as a cache invalidator. 

## How to install

Mobile menu is still a work in progress. To test this issue, some tricks are needed:
* Get the feature branch from platfom config: `composer require drupal/helfi_platform_config:dev-UHF-6038-menu-data-syncing`
* Enable global menu feature: `drush en helfi_navigation -y`
* (Get this branch: `composer require drupal/hdbt:dev-UHF-6198-mobile-menu-caching)` Change line 259 of `themes/contrib/hdbt/src/js/nav-global.js` to `const MENU = mockmenu;`
* Checkout branch. Run `nvm use; npm i; npm run build`
* Clear cache `drush cr`

## How to test
* Load a page and scale down to mobile view. You should see the hamburger menu in the top right corner of the page. Click it.
* After the menu has loaded, check that it's been cached: `const x = JSON.parse(localStorage.getItem('hdbt-mobile-menu'))` in the console. The returned object should hold the menu data + a timestamp.
* Check the timestamp: `x.timestamp` to print it out.
* Overwrite the timestamp with and hour old one: `let y = x; y.timestamp = new Date().getTime() - 60 * 60 * 1000; localstrorage.setItem('hdbt-mobile-menu', JSON.stringify(y))`;. Take note of the timestamp you just created.
* Reload the page and open the mobile menu again. Print out the cache timestamp again (JSON.parse(localStorage.getItem('hdbt-mobile-menu')).timestamp). The hour old timestamp should have been updated with an up-to-date one.


## Designers review



## Other PRs


* Link to other PR
